### ISSUE TITLE
Number formatting

### ISSUE THREAD

**Issue by himamis** [a14n/dart-decimal#35]
It would be nice to have a number formatting package that works with the decimal (rational) packages. 
I think it would be possible to modify the library code to make it compatible with intl/NumberFormat formatter.
Do you have any plans to support this?

**Comment by a14n** [comment]
No, I haven't thought about this. 

> I think it would be possible to modify the library code to make it compatible with intl/NumberFormat formatter.

What should be done to make it work?

**Comment by himamis** [comment]
The NumberFormat package is implemented expecting a dynamic type. It uses standard number like operators to format the number. But sometimes, it passes numbers to these operators, which are not expected by the decimal (rational) package.
I was able to make the package work by replacing the times, division, truncating division and remainder operators parameter with a dynamic type, i.e.:
```
-  Rational remainder(Rational other) => this - (this ~/ other) * other;
+  Rational remainder(dynamic other) => this - (this ~/ other) * other;
```
```
-  Rational operator ~/(Rational other) => (this / other).truncate();
+  Rational operator ~/(dynamic other) => (this / Rational.tryParse(other)).truncate();
```
```
-  Rational operator /(Rational other) =>
-      Rational(numerator * other.denominator, denominator * other.numerator);
+  Rational operator /(dynamic other) {
+    final otherRational = Rational.tryParse(other);
+    return Rational(numerator * otherRational.denominator, denominator * otherRational.numerator);
+  }
```
```
-  Rational operator *(Rational other) =>
-      Rational(numerator * other.numerator, denominator * other.denominator);
+  Rational operator *(dynamic other) {
+    final otherRational = Rational.tryParse(other);
+    return Rational(numerator * otherRational.numerator, denominator * otherRational.denominator);
+  }
```
and adding the factory method
```
+  factory Rational.tryParse(dynamic other) {
+    if (other is Rational) {
+      return other;
+    } else if (other is int) {
+      return Rational.fromInt(other);
+    }
+    return null;
+  }
```
Similar to `double.tryParse`.

**Comment by a14n** [comment]
I'm not a big fan of using dynamic for parameters because it could hide bugs until runtime. However we could create an adapter and use it `NumberFormat.format(DecimalAdapter(decimal))`.

Finally we should be very careful about what the values of these parameter are and that those values have no precision lost.

**Comment by himamis** [comment]
Sounds good. I will look into this a bit later and create a pull request.

**Comment by kentcb** [comment]
Arrived here looking for the same thing. I ended up just using `toDouble()` and passing that to `NumberFormat`, although I'm aware this is less than ideal.

**Comment by himamis** [comment]
I did not have the time to create a pull request, but I have been using this since:
```
import 'package:decimal/decimal.dart';
import 'package:flutter/foundation.dart';

@immutable
class DecimalAdapter {
  final Decimal decimal;

  DecimalAdapter(this.decimal);

  factory DecimalAdapter.tryParse(dynamic number) {
    if (number is DecimalAdapter) {
      return number;
    } else if (number is Decimal) {
      return DecimalAdapter(number);
    } else if (number is int) {
      return DecimalAdapter(Decimal.fromInt(number));
    }
    return DecimalAdapter(Decimal.parse(number.toString()));
  }

  bool get isNaN => decimal.isNaN;

  bool get isNegative => decimal.isNegative;

  DecimalAdapter operator ~/(dynamic other) =>
      DecimalAdapter(decimal ~/ DecimalAdapter.tryParse(other).decimal);

  bool get isInfinite => decimal.isInfinite;

  DecimalAdapter abs() => DecimalAdapter(decimal.abs());

  DecimalAdapter operator +(dynamic other) =>
      DecimalAdapter(decimal + DecimalAdapter.tryParse(other).decimal);

  DecimalAdapter operator -(dynamic other) =>
      DecimalAdapter(decimal - DecimalAdapter.tryParse(other).decimal);

  DecimalAdapter operator *(dynamic other) =>
      DecimalAdapter(decimal * DecimalAdapter.tryParse(other).decimal);

  DecimalAdapter operator /(dynamic other) =>
      DecimalAdapter(decimal / DecimalAdapter.tryParse(other).decimal);

  DecimalAdapter remainder(dynamic other) =>
      DecimalAdapter(decimal.remainder(DecimalAdapter.tryParse(other).decimal));

  int toInt() => decimal.toInt();

  @override
  bool operator ==(Object other) => decimal == DecimalAdapter.tryParse(other).decimal;

  @override
  String toString() => decimal.toString();

  double toDouble() => decimal.toDouble();

  DecimalAdapter round() => DecimalAdapter(decimal.round());

  @override
  int get hashCode => decimal.hashCode;
}
```

To use it, call `NumberFormat().format(DecimalAdapter(decimal))`

**Comment by a14n** [comment]
Fixed by 203f28cd6ed86d17454093409465376a113eb2b9 available in decimal-1.2.0.

You have to use `DecimalIntl` as mentioned in [README](https://github.com/a14n/dart-decimal/tree/272d5835b82f688537a43c1c494f37fcf61b9e22#formatting-with-intl)
